### PR TITLE
Added point pool restriction option in world generation settings

### PR DESCRIPTION
**Pull request by phantom-voltage**
In reference to #21164, I changed the no_freeform mod to be a world gen option, which allows for:

- All point pools
- Multi-pool only
- Single-pool only
- No freeform pool

This is made to address balance concerns for the shared-map mode.

To address @DangerNoodle 's concern about changing the stat values, in shared map mode the debug options are not available to users without the admin flag. So this should suffice to control user character generation.

### CODE REVIEW

**Comment by Coolthulhu** [comment]
> Single-pool only

This option is redundant. Multi-pool can never be easier than single-pool.

**Comment by Coolthulhu** [comment]
There should be 3 options:
* Unlimited (freeform allowed)
* No freeform
* Multi-pool only

File: src/newcharacter.cpp [hunk 1]
```diff
@@ -696,20 +696,54 @@ tab_direction set_points( WINDOW *w, player *, points_left &points )
...
+
+    const point_limit_tuple one_pool = std::make_tuple( points_left::ONE_POOL, _( "Single pool" ),
+                         _( "Stats, traits and skills share a single point pool." ) );
+
+    const point_limit_tuple freeform = std::make_tuple( points_left::FREEFORM, _( "Freeform" ),
+                         _( "No point limits are enforced" ) );
+
+
+    if( point_pool == "multiple" ) {
+
```
Comment (Coolthulhu) [inline src/newcharacter.cpp]: There is a lot of extra spacing here.
Don't add double newline after opened braces, it stretches the whole thing heavily.

File: src/newcharacter.cpp [hunk 2]
```diff
@@ -696,20 +696,54 @@ tab_direction set_points( WINDOW *w, player *, points_left &points )
...
+    const point_limit_tuple freeform = std::make_tuple( points_left::FREEFORM, _( "Freeform" ),
+                         _( "No point limits are enforced" ) );
+
+
+    if( point_pool == "multiple" ) {
+    
+        opts = {{multi_pool}};
+    }
+
+    else if( point_pool == "single" ) {
```
Comment (Coolthulhu) [inline src/newcharacter.cpp]: Should be on the same line as closing brace:
```} else if( x ) {```

File: src/newcharacter.cpp [hunk 3]
```diff
@@ -696,20 +696,54 @@ tab_direction set_points( WINDOW *w, player *, points_left &points )
...
+    
+        opts={{one_pool}};
+
+        // Set selected point pool to single, otherwise multipool is chosen by default 
+        points.limit = std::get<0>((opts[0]));
```
Comment (Coolthulhu) [inline src/newcharacter.cpp]: You can yank it out, after the `if/else` ladder, so that the `if/else` only sets `opts` - that way it will be more readable.

**Comment by phantom-voltage** [comment]
Thank you, @Coolthulhu. I'll make the requested changes. I also changed the final else statement to default to ALL if the world was created without this option.

File: src/newcharacter.cpp (hunk 3 above)
Comment (Coolthulhu) [inline src/newcharacter.cpp]: I mean, it can be unconditional

Comment (phantom-voltage) [inline src/newcharacter.cpp]: That's how I had it originally, setting the point limit after, however I thought it was redundant when points.limit defaults to MULTI_POOL in all other cases.

In the future, should legibility be prioritized?

**Comment by phantom-voltage** [comment]
I agree with the 3 options, my only concern is the implication of "unlimited." At a glance, this sounds like it's freeform. Is the wording "all" acceptable?

**Comment by Coolthulhu** [comment]
> Is the wording "all" acceptable?

How about "any"? "All" suggests all at once.

File: src/newcharacter.cpp (hunk 3 above)
Comment (Coolthulhu) [inline src/newcharacter.cpp]: > In the future, should legibility be prioritized?

It's more about the good habit of keeping data and procedures separates.
Think of it this way: the `if/else` ladder is essentially a dictionary (more simply: an array addressed with strings).

When data and procedures are separate, maintenance is easier. I have changed multiple `if/else` ladders into pure data (jsons or just arrays/dictionaries) and it was always the extra functions causing the most trouble. The rest could easily be handled with search+replace.

It's not necessary, but it's a good habit that makes everyone's jobs easier.

File: src/newcharacter.cpp [hunk 4]
```diff
@@ -696,18 +696,32 @@ tab_direction set_points( WINDOW *w, player *, points_left &points )
...
+    const point_limit_tuple one_pool = std::make_tuple( points_left::ONE_POOL, _( "Single pool" ),
+                         _( "Stats, traits and skills share a single point pool." ) );
+
+    const point_limit_tuple freeform = std::make_tuple( points_left::FREEFORM, _( "Freeform" ),
+                         _( "No point limits are enforced" ) );
 
+
+    if( point_pool == "multiple" ) {
+        opts = {{multi_pool}};
+    } else if( point_pool == "no_freeform") {
```
Comment (Coolthulhu) [inline src/newcharacter.cpp]: Spacing:
You have it:
```
if( x) {
y={{z}};
```
Should be:
```
if( x ) {
y = {{ z }};
```

That is, parentheses of `if` should be padded with spaces from both sides, arrays should be padded with spaces from both sides.

File: src/options.cpp [hunk 5]
```diff
@@ -1479,6 +1479,16 @@ void options_manager::init()
 
     mOptionsSort["world_default"]++;
 
+    optionNames["any"] = _("Any");
+    optionNames["multiple"] = _("Multiple");
```
Comment (Coolthulhu) [inline src/options.cpp]: "Multiple" sounds weird here, as if you could pick more than one.
"Multi-pool only" would be better.

File: src/options.cpp [hunk 6]
```diff
@@ -1479,6 +1479,16 @@ void options_manager::init()
...
+    optionNames["no_freeform"] = _("No Freeform");
+    add("CHARACTER_POINT_POOLS", "world_default", _("Character point pools"),
+        _("Allowable point pools for character generation."),
```
Comment (Coolthulhu) [inline src/options.cpp]: "Allowable" -> "Allowed"

**Comment by phantom-voltage** [comment]
I was looking at the status of the pull request, and once I updated my branch, I saw that the travis build failed this time with clang++. The failure was related to a creature_test.cpp. As far as I know, my proposed changes are out of this scope.

I'm not sure how or why this failed. Could someone point me in the right direction to fix this?

**Comment by DangerNoodle** [comment]
Had the same issue earlier, seems that test occasionally fails without any fault on the pull request opener.

**Comment by phantom-voltage** [comment]
Are things good to go on my end? Is there anything else I need to address or change?

**Pull request merged**